azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -852,14 +728,14 @@ def do_retry():
             )
 
             # if we don't keep track of this op, it might get collected.
+            op.uncomplete()
             self.ops_waiting_to_retry.append(op)
             op.retry_timer = Timer(self.retry_intervals[type(op)], do_retry)
             op.retry_timer.start()
 
         else:
             if op.retry_timer:
                 op.retry_timer = None
```

Comment:
>retry_timer [](start = 19, length = 11)

possible unrelated bug: op.retry_timer.cancel() should be called.

Reply:
+1. 

---
In reply to: [345999596](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r345999596) [](ancestors = 345999596)

Reply:
I'll attempt to fix this in PR #2

---
In reply to: [345999596](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r345999596) [](ancestors = 345999596)

Reply:
I've made the change (and also in the `do_retry` callback slightly above), but it seems like there are no tests for this entire code branch. The `do_retry` callback has tests that I've adjusted, but I can't seem to find them for this else block. Am I just missing something?

---
In reply to: [345999596](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r345999596) [](ancestors = 345999596)

Reply:
You're correct.  I have a test to make sure the callback gets called but we don't have a test to make sure the timer is cancelled:
@pytest.mark.it("Calls the original callback with success when the retried op succeeds")
In truth, I'm having trouble figuring out how this code could even be hit because it means that op completes while the op is waiting to be retried (which means it must be completed already).  Can you mark this with a BKTODO please?  I'm reworking this code path and I need to make sure I think this through.  


---
In reply to: [347059768](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r347059768) [](ancestors = 347059768,345999596)